staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/controller_ref.go

```diff
@@ -31,9 +31,10 @@ func IsControlledBy(obj Object, owner Object) bool {
 
 // GetControllerOf returns a pointer to a copy of the controllerRef if controllee has a controller
 func GetControllerOf(controllee Object) *OwnerReference {
-	for _, ref := range controllee.GetOwnerReferences() {
-		if ref.Controller != nil && *ref.Controller {
-			return &ref
+	refs := controllee.GetOwnerReferences()
+	for i := range refs {
+		if refs[i].Controller != nil && *refs[i].Controller {
+			return &refs[i]
```

Comment:
this change seems to go against the documented behavior of the method. it seems likely clients could be depending on getting a copy rather than modifying the reference in-place in the object.

Reply:
What if we do:
```
return &(*refs[i])
```

Then we at least avoid copying all refs if there are more.

Reply:
that seems ok. does the benchmark still show improvement with that?

Reply:
The current benchmark will, because there are many owner refs. In practice it won't be super helpful, because in majority of cases we have a single owner ref. But it also won't be worse (i.e. we will be copying exactly one item, instead of all).

Reply:
> we will be copying exactly one item, instead of all

we were iterating, but only allocating a single item, right?

Reply:
Thanks, Jordan. That's a really great point. 

I changed the method to return the copy to fulfill the contract from the method doc. Unfortunately, this reduced the performance improvement from 10x to 10%, but it's still something. 
In addition I added a new flavor of this method that doesn't copy and used it in the IsControllerdBy method. This is a really low hanging fruit that should give us some improvement. We can change more callers in future, if there is a need.

Reply:
I also updated the benchmark to be more realistic.

Reply:
sounds good. Looks like there are a few other call sites in common controller utils we can easily update:
* realHistory#ListControllerRevisions could switch to use IsControlledBy
* realHistory#AdoptControllerRevision could switch to GetControllerOfNoCopy
* BaseControllerRefManager#ClaimObject could switch to GetControllerOfNoCopy

Reply:
+1

let's change it in this PR given you're touching this code already

Reply:
Changed these 3 places as well.